<a href="https://colab.research.google.com/github/kipkii/kipki-s-/blob/main/%ED%86%B5%EA%B3%84%EC%9A%94%EC%95%BD.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

04-1 기초 통계

In [ ]:
import gdown
gdown.download('https://bit.ly/3736JW1','ns_book6.csv',quiet=False)

Downloading...
From: https://bit.ly/3736JW1
To: /content/ns_book6.csv
100%|██████████| 55.0M/55.0M [00:01<00:00, 46.2MB/s]


'ns_book6.csv'

In [ ]:
import pandas as pd

ns_book6 = pd.read_csv('ns_book6.csv', low_memory=False)
ns_book6.head()

,번호,도서명,저자,출판사,발행년도,ISBN,세트 ISBN,부가기호,권,주제분류번호,도서권수,대출건수,등록일자
0,1,인공지능과 흙,김동훈 지음,민음사,2021.0,9788937444319,NaN,NaN,NaN,NaN,1,0,2021-03-19
1,2,가짜 행복 권하는 사회,김태형 지음,갈매나무,2021.0,9791190123969,NaN,NaN,NaN,NaN,1,0,2021-03-19
2,3,나도 한 문장 잘 쓰면 바랄 게 없겠네,김선영 지음,블랙피쉬,2021.0,9788968332982,NaN,NaN,NaN,NaN,1,0,2021-03-19
3,4,예루살렘 해변,"이도 게펜 지음, 임재희 옮김",문학세계사,2021.0,9788970759906,NaN,NaN,NaN,NaN,1,0,2021-03-19
4,5,김성곤의 중국한시기행 : 장강·황하 편,김성곤 지음,김영사,2021.0,9788934990833,NaN,NaN,NaN,NaN,1,0,2021-03-19


In [ ]:
ns_book6.describe()

,번호,발행년도,도서권수,대출건수
count,379976.000000,379976.000000,379976.000000,379976.000000
mean,201726.332847,2008.516306,1.135874,11.504629
std,115836.454596,8.780529,0.483343,19.241926
min,1.000000,1947.000000,0.000000,0.000000
25%,102202.750000,2003.000000,1.000000,2.000000
50%,203179.500000,2009.000000,1.000000,6.000000
75%,301630.250000,2015.000000,1.000000,14.000000
max,401681.000000,2650.000000,40.000000,1765.000000


In [ ]:
sum(ns_book6['도서권수'] == 0)

TypeError: 'int' object is not callable

In [ ]:
ns_book7 = ns_book6[ns_book6['도서권수'] > 0]

In [ ]:
ns_book7.describe(percentiles=[0.3, 0.6, 0.9])

,번호,발행년도,도서권수,대출건수
count,376770.000000,376770.000000,376770.000000,376770.000000
mean,202977.476649,2008.460076,1.145540,11.593439
std,115298.245784,8.773148,0.473853,19.279409
min,1.000000,1947.000000,1.000000,0.000000
30%,124649.700000,2004.000000,1.000000,2.000000
50%,204550.500000,2009.000000,1.000000,6.000000
60%,243537.400000,2011.000000,1.000000,8.000000
90%,361341.100000,2018.000000,2.000000,28.000000
max,401681.000000,2650.000000,40.000000,1765.000000


In [ ]:
ns_book7.describe(include = 'object')

,도서명,저자,출판사,ISBN,세트 ISBN,부가기호,권,주제분류번호,등록일자
count,376770,376770,376770,376770,55866,308252,61793,359792,376770
unique,336408,248850,21875,350810,14875,17,834,12467,4562
top,승정원일기,세종대왕기념사업회 [편],문학동네,9788937430299,9788937460005,0,1,813.6,1970-01-01
freq,250,303,4410,206,702,158235,13282,14816,28185


count는 누락된 값 제외 데이터 개수,
unique 행은 고유한 값의 개수
top행은 가장많이 등장하는 값 나타냄
freq행은 top행에 등장하는 항목의 빈도수
실행결과에 따르면 도서명 열에서 가장많이 등장하는 데이터는 승정원일기로 250회!
판다스의 describe() 메서드는 이처럼 다양한 통계량을 간편 요약 확인 가능
평균, 중앙갑등 개발 통계량에대해 조금더 알아보자!

In [ ]:
x = [10,20,30]
result = 0
for i in range(3):
  re += x[i]
print("평균 :", sum/len(x))

평균 : 20.0


In [ ]:
ns_book7['대출건수'].mean()

np.float64(11.593438968070707)

판다스 데이터프레임과 시리즈 객체는 평균을 계산하는 mean()메서드를 제공합니다.
평균 대출 건수는 약 11.6으로 describe() 메서드를 출력해 본것과 같다.

In [ ]:
ns_book7['대출건수'].median()

6.0

중앙값은 말그대로 전체 데이터를 순서대로 늘어 놓았을 떄 주앙에 위차한 값이다.
50% 위치가 중앙값이다!

In [ ]:
temp_df = pd.DataFrame([1,2,3,4])
temp_df.median(ㅗ)

,0
0,2.5


데이터 개수가 홀수 일떄는 중앙값을 결정하기 쉽다. 정확히 가운데 놓인 값을 찾으면 된다. 하지만 짝수일때는 가운데 두개를 평균하여 중앙값을 결정한다.

## **중복값 제거하고 중앙값 구하기**

In [ ]:
ns_book7['대출건수'].drop_duplicates().median()

183.0

중복된 값없애라!

최대, 최소

In [ ]:
ns_book7['대출건수'].min()

0

In [ ]:
ns_book7['대출건수'].max()

1765

# 분위수 (quartile)



분위수는 데이터를 순서대로 늘어 놓았을 떄 이를 균등한 간격으로 나누는 기준점 입니다. 예를 들어 이분위수는 전체데이터를 두구간으로 나누기 때문에 중앙값에 해당
따라서 사분위수는 3개가 나오고, 각각 25, 50 75에 해당
25에 대항하는 값을 1사분위수, 중앙값을 2 사분위수, 75값을 3사분위 수라고도 부릅니다.

In [ ]:
ns_book6['대출건수'].quantile(0.25)

np.float64(2.0)

In [ ]:
ns_book6['대출건수'].quantile([0.25,0.5,0.75])

,대출건수
0.25,2.0
0.50,6.0
0.75,14.0


In [ ]:
pd.Series([1,2,3,4,5]).quantile(0.9)

np.float64(4.6)

In [ ]:
pd.Series([1,2,3,4,5]).quantile(0.9, interpolation='midpoint')

np.float64(4.5)

In [ ]:
pd.Series([1,2,3,4,5]).quantile(0.9, interpolation='nearest')

np.int64(5)

두지점사이에 놓인 특정 위치 값을 구하는 방법을 보간(interploation)이라고 부른다
interpolation매개변수는 linear외에도 여러가지 다른 보간 방식을 제공
midpoint는 분위수에 상관없이 무조건 두 수 사이의 중앙값을 사용한다.
nearest는 두수중에서 가까운 값을 사용한다.

In [ ]:
borrow_10_flag = ns_book7['대출건수'] <10

In [ ]:
borrow_10_flag.mean()

np.float64(0.6402712530190833)

In [ ]:
ns_book7['대출건수'].quantile(0.65)

np.float64(10.0)

ns_book7.var()

In [ ]:
ns_book7['대출건수'].var()

371.6956304306922

In [ ]:
ns_book7['대출건수'].std()

19.27940949382766

평균보다 훨씬 큰 대출건수가 있기 때문에 표준편차가 평균보다 많이 클 수 있음을 알 수 있다.

판다스의 std 메서드를 수식으로 구현하려면 샘플 개수를 n-1로 나눠야함
자유도 평균이 정해지는 순간 자유도는 n-1

In [ ]:
import numpy as np
diff = ns_book7['대출건수'] - ns_book7['대출건수'].mean()
np.sqrt(np.sum(diff**2) / (len(ns_book7)-1))


np.float64(19.279409493785508)

# 최빈값 구하기


가장 많이 등장하는 값! mode()메서드를 이용한다

In [ ]:
ns_book7['도서명'].mode()

,도서명
0,승정원일기


In [ ]:
ns_book7['발행년도'].mode()

,발행년도
0,2012.0


In [ ]:
ns_book7.mean(numeric_only=True)

,0
번호,202977.476649
발행년도,2008.460076
도서권수,1.145540
대출건수,11.593439


열에 대한 최빈값!

In [ ]:
ns_book7.loc[:,'도서명':].mode()

,도서명,저자,출판사,발행년도,ISBN,세트 ISBN,부가기호,권,주제분류번호,도서권수,대출건수,등록일자
0,승정원일기,세종대왕기념사업회 [편],문학동네,2012.0,9788937430299,9788937460005,0,1,813.6,1,0,1970-01-01


In [ ]:
ns_book7.to_csv('ns_book7.csv', index = False)

##넘파이의 기술 통계 함수






In [ ]:
import numpy as np
np.mean(ns_book7['대출건수'])

np.float64(11.593438968070707)

In [ ]:
np.average(ns_book7['대출건수'], weights = 1/ ns_book7['도서권수'])

np.float64(10.543612175385386)

In [ ]:
np.mean(ns_book7['대출건수']/ns_book7['도서권수'])

np.float64(9.873029861445774)

In [ ]:
np.mean(ns_book7['대출건수'].sum()/ns_book7['도서권수'].sum())

np.float64(10.120503701300958)

In [ ]:
np.median(ns_book7['대출건수'])

np.float64(6.0)

In [ ]:
np.min(ns_book7['대출건수'])

0

In [ ]:
np.max(ns_book7['대출건수'])

1765

In [ ]:
np.quantile(ns_book7['대출건수'],[0.25, 0.5, 0.75])

array([ 2.,  6., 14.])

In [ ]:
np.var(ns_book7['대출건수'])

371.694643898775

In [ ]:
ns_book7['대출건수'].var()

371.6956304306922

소수점 셋째 자리부터 다르다..!

#n-1 자유도

판다스는 분산을 할 때 분모가 n이 아닌 n-1
표본집단을 사용해서 모집단의 특징을 추정
넘파이의 var()함수는 n으로 나누기 때문에 판다스와 결과가 다름
단순 일련의 데이터에 대한 분산을 계산
판다스와 넘파이의 ddof 값을 각각 0,1로 지정하여 분산 결과 확인

In [ ]:
np.std(ns_book7['대출건수'])

19.279383908693113

return_counts 매개변수를 기본값 False에서 True로 바꿔 주면 고유한 값의 등장 횟수도 반환
unique( ) 함수를 호출해 고유한 값(values)과 등장 횟수(counts) 배열을 획득



In [ ]:
values, counts = np.unique(ns_book7['도서명'],return_counts = True)

In [ ]:
max_idx = np.argmax(counts)

In [ ]:
values[max_idx]

'승정원일기'